File: mobile/app/src/main/java/org/neuronavx/logger/nav/Signals.kt
package org.neuronavx.logger.nav

import kotlin.math.abs
import kotlin.math.hypot
import kotlin.math.sqrt

/**
 * Derived inertial signals, in the conventions the desktop pipeline validated empirically
 * against IO-VNBD (see `src/neuronav/calibration/signals.py` and `docs/data_audit.md`).
 *
 * The two that matter and are easy to get wrong:
 *  - the reported GRAVITY vector points UP (it is the specific-force direction), so
 *    linear acceleration is `accel - gravity`, and the ORIENTATION/rotation-vector
 *    channels are never used -- they were shown to be inconsistent with gravity.
 *  - a right-handed rotation about "up" is counter-clockwise while compass heading grows
 *    clockwise, which is why heading rate carries a negation on the desktop side. Here the
 *    yaw rate comes from a single calibrated gyro channel instead, because the correct
 *    channel differs per device and matches no axis label.
 *
 * Everything is written for one sample at a time: this runs inside a sensor callback, not
 * over an array.
 */
object Signals {

    const val GRAVITY_MAGNITUDE = 9.80665

    /** Unit vector along the device-frame gravity ("up") direction. */
    fun gravityUnit(g: DoubleArray, out: DoubleArray) {
        val norm = sqrt(g[0] * g[0] + g[1] * g[1] + g[2] * g[2]).coerceAtLeast(1e-9)
        out[0] = g[0] / norm
        out[1] = g[1] / norm
        out[2] = g[2] / norm
    }

    /**
     * Two orthonormal device-frame vectors spanning the local horizontal plane.
     *
     * Built by projecting the device X axis onto the plane perpendicular to gravity, and
     * falling back to Y when X is near-vertical -- otherwise the projection degenerates and
     * the basis spins arbitrarily, which would make the alignment angle meaningless.
     */
    fun horizontalBasis(up: DoubleArray, e1: DoubleArray, e2: DoubleArray) {
        val useY = abs(up[0]) > 0.9
        val refX = if (useY) 0.0 else 1.0
        val refY = if (useY) 1.0 else 0.0

        val dot = refX * up[0] + refY * up[1]
        e1[0] = refX - up[0] * dot
        e1[1] = refY - up[1] * dot
        e1[2] = 0.0 - up[2] * dot
        val norm = sqrt(e1[0] * e1[0] + e1[1] * e1[1] + e1[2] * e1[2]).coerceAtLeast(1e-9)
        e1[0] /= norm; e1[1] /= norm; e1[2] /= norm

        // e2 = up x e1
        e2[0] = up[1] * e1[2] - up[2] * e1[1]
        e2[1] = up[2] * e1[0] - up[0] * e1[2]
        e2[2] = up[0] * e1[1] - up[1] * e1[0]
    }

    /**
     * Horizontal components of the linear (gravity-removed) acceleration, in the (e1, e2)
     * basis. Returns into [out] as (h1, h2).
     */
    fun horizontalAcceleration(accel: DoubleArray, gravity: DoubleArray, out: DoubleArray) {
        val up = DoubleArray(3)
        val e1 = DoubleArray(3)
        val e2 = DoubleArray(3)
        gravityUnit(gravity, up)
        horizontalBasis(up, e1, e2)

        val lx = accel[0] - gravity[0]
        val ly = accel[1] - gravity[1]
        val lz = accel[2] - gravity[2]
        out[0] = lx * e1[0] + ly * e1[1] + lz * e1[2]
        out[1] = lx * e2[0] + ly * e2[1] + lz * e2[2]
    }

    /** Specific force along gravity, minus 1 g -- the vertical feature channel. */
    fun verticalAcceleration(accel: DoubleArray, gravity: DoubleArray): Double {
        val up = DoubleArray(3)
        gravityUnit(gravity, up)
        return accel[0] * up[0] + accel[1] * up[1] + accel[2] * up[2] - GRAVITY_MAGNITUDE
    }

    fun magnitude(v: DoubleArray): Double = sqrt(v[0] * v[0] + v[1] * v[1] + v[2] * v[2])

    /**
     * Vehicle heading rate from Android-native sensors, positive clockwise from North.
     *
     * Android's gyroscope and gravity sensors use the same device coordinate frame. A
     * vehicle yaw is therefore the angular-rate component about local vertical; the minus
     * sign converts Android's right-handed counter-clockwise convention to compass heading.
     * Unlike a frozen single-axis gain, this remains correct when the phone's pitch/roll in
     * its cradle changes while driving.
     *
     * IO-VNBD replay columns do not share this axis contract, so replay navigation retains
     * its calibrated-channel path in [Navigator].
     */
    fun gravityProjectedHeadingRate(gyro: DoubleArray, gravity: DoubleArray): Double {
        val norm = magnitude(gravity)
        if (!norm.isFinite() || norm < 1.0) return Double.NaN
        return -(gyro[0] * gravity[0] + gyro[1] * gravity[1] + gyro[2] * gravity[2]) / norm
    }

    fun hypot2(x: Double, y: Double): Double = hypot(x, y)

    /** Wrap to (-pi, pi]. */
    fun wrapAngle(a: Double): Double {
        val twoPi = 2.0 * Math.PI
        var x = (a + Math.PI) % twoPi
        if (x < 0) x += twoPi
        return x - Math.PI
    }
}

File: src/neuronav/calibration/signals.py
"""Derived inertial signals, using conventions validated empirically against IO-VNBD.

Validation summary (see docs/data_audit.md):

* The GRAVITY columns agree with the accelerometer to within 0.02-0.44 deg, so the
  gravity vector is the trustworthy attitude reference. The ORIENTATION columns are
  NOT self-consistent with it and are never used here.
* The reported gravity vector points UP (it is the specific-force direction: +Z when
  the phone lies flat, screen up).
* Heading rate is the angular-rate component along that up axis, negated, because a
  right-handed rotation about "up" is counter-clockwise while compass heading grows
  clockwise. Regressing integrated yaw against GNSS course change over 5 s baselines
  gave slopes of -0.974 / -0.927 / -0.963 on segments A5 / A7 / A8.
"""
import numpy as np
import pandas as pd

GRAVITY_MAGNITUDE = 9.80665


def gravity_unit(df: pd.DataFrame) -> np.ndarray:
    """(N,3) unit vector along the device-frame gravity ('up') direction."""
    g = np.column_stack([df["grav_x"], df["grav_y"], df["grav_z"]])
    norm = np.linalg.norm(g, axis=1, keepdims=True)
    return g / np.maximum(norm, 1e-9)


def angular_rate(df: pd.DataFrame) -> np.ndarray:
    """(N,3) device-frame angular rate in rad/s, in the CSV's column order."""
    return np.column_stack([df["gx"], df["gy"], df["gz"]])


def heading_rate(df: pd.DataFrame) -> np.ndarray:
    """(N,) vehicle heading rate in rad/s, positive clockwise (compass sense)."""
    return -np.sum(angular_rate(df) * gravity_unit(df), axis=1)


def linear_acceleration(df: pd.DataFrame) -> np.ndarray:
    """(N,3) device-frame acceleration with gravity removed, m/s^2."""
    a = np.column_stack([df["ax"], df["ay"], df["az"]])
    g = np.column_stack([df["grav_x"], df["grav_y"], df["grav_z"]])
    return a - g


def horizontal_basis(df: pd.DataFrame) -> tuple[np.ndarray, np.ndarray]:
    """Two orthonormal device-frame vectors spanning the local horizontal plane.

    Built by projecting the device X axis onto the plane perpendicular to gravity,
    falling back to device Y where the phone is held such that X is near-vertical.
    """
    up = gravity_unit(df)
    ref = np.tile(np.array([1.0, 0.0, 0.0]), (len(up), 1))
    degenerate = np.abs(np.sum(ref * up, axis=1)) > 0.9
    ref[degenerate] = np.array([0.0, 1.0, 0.0])

    e1 = ref - up * np.sum(ref * up, axis=1, keepdims=True)
    e1 /= np.maximum(np.linalg.norm(e1, axis=1, keepdims=True), 1e-9)
    e2 = np.cross(up, e1)
    return e1, e2


def horizontal_acceleration(df: pd.DataFrame) -> np.ndarray:
    """(N,2) linear acceleration resolved in the (e1, e2) horizontal basis, m/s^2."""
    lin = linear_acceleration(df)
    e1, e2 = horizontal_basis(df)
    return np.column_stack([np.sum(lin * e1, axis=1), np.sum(lin * e2, axis=1)])


def tilt_angles(df: pd.DataFrame) -> tuple[np.ndarray, np.ndarray]:
    """(pitch, roll) in radians derived from gravity, Android sign convention."""
    up = gravity_unit(df)
    pitch = np.arcsin(np.clip(-up[:, 1], -1.0, 1.0))
    roll = np.arctan2(-up[:, 0], up[:, 2])
    return pitch, roll
